### ISSUE TITLE
integrating into nixos-aarch64-images?

### ISSUE THREAD

**Issue by Mic92** [tmountain/arch-nanopct4#1]
We already have some logic for rocket chips in https://github.com/Mic92/nixos-aarch64-images
So all we would need is a fetchurl to download uboot.
I cannot test the image since I don't own the hardware, but I would happily accept a pull request. This would make it easier for people to flash images.

**Comment by Mic92** [comment]
Also the image you referenced in the wiki is quite old 18.09. An update would be great.

**Comment by tmountain** [comment]
@Mic92 Thanks for reaching out. I would like to work on this with you. I am in the process of trying to figure out how whether the NanoPC-T4 can make use of the upstream u-boot repo without modification. There is an open pull request where that is currently being investigated:

https://github.com/NixOS/nixpkgs/pull/109423

Once those questions are resolved, I can submit a PR and assist with testing. The build image in the wiki has been updated, and I'm also working on finishing up that guide. It may change depending upon how the u-boot issue resolves.

**Comment by Mic92** [comment]
Ok. I would also accept uboot patches or prebuild uboot blobs if that helps to make it boot, but ideally in a way that can be build from x86_64.

**Comment by tmountain** [comment]
Got it. That certainly makes it easier. I will follow-up with a PR soon.

**Comment by tmountain** [comment]
@Mic92 I have created a repo dedicated to building u-boot for the NanoPC-T4. Please see the README.md for details of my findings. If you would like, I can work towards a PR using this repo as a basis, or alternatively, you can incorporate this work into your repository. Let me know, and we can proceed.

https://github.com/tmountain/u-boot-nanopct4

**Comment by Mic92** [comment]
Yeah. Let's integrate it. The bootloader needs to go here: https://github.com/Mic92/nixos-aarch64-images/blob/2ddb57c544cc49f8ddb18823444833787e99e00a/default.nix#L17
As it is a rockchip too, I assume the existing abstraction should just work?

**Comment by tmountain** [comment]
I'm pretty close to having a working PR, but I'm not sure how to compile/test. I'm trying to run nix-build on my x86 machine, and I get the following error:

```
error: a 'aarch64-linux' with features {} is required to build '/nix/store/0pafg292z6b9ynj0d6a2dbjj3b11d7d4-builder.pl.drv', but I am a 'x86_64-linux' with features {benchmark, big-parallel, nixos-test}
```

I'm not sure how to cross-compile for a `default.nix` file. Can you tell me how to build this on x86?

**Comment by Mic92** [comment]
You can use `pkgsCross.aarch64-multiplatform`, for example:

```
$ nix-build '<nixpkgs>' -A pkgsCross.aarch64-multiplatform.hello
```

**Comment by tmountain** [comment]
That command looks like it applies to packages that are already in &lt;nixpkgs&gt;. In the case of your project, it's in its own repo. Can you provide a specific example as to how I would build the rock64 target using the syntax above?

```
# nix-build '<nixpkgs>' -A pkgsCross.aarch64-multiplatform.rock64
error: attribute 'rock64' in selection path 'pkgsCross.aarch64-multiplatform.rock64' not found
```

Thanks!

**Comment by Mic92** [comment]
> That command looks like it applies to packages that are already in <nixpkgs>. In the case of your project, it's in its own repo. Can you provide a specific example as to how I would build the rock64 target using the syntax above?
> 
> ```
> # nix-build '<nixpkgs>' -A pkgsCross.aarch64-multiplatform.rock64
> error: attribute 'rock64' in selection path 'pkgsCross.aarch64-multiplatform.rock64' not found
> ```
> 
> Thanks!
```
pkgsCross.aarch64-multiplatform.callPackage ./uboot.nix {}
```

**Comment by tmountain** [comment]
Got it. It looks like it still does not want to build (fails on arm-trusted-firmware):

```
# nix-build -E 'with import <nixpkgs> {}; pkgsCross.aarch64-multiplatform.callPackage ./default.nix {}'
...
error: a 'aarch64-linux' with features {} is required to build '/nix/store/mr166g96ync4hlnm4046y878jnp1yyv0-arm-trusted-firmware-rk3399-2.4.drv', but I am a 'x86_64-linux' with features {benchmark, big-parallel, nixos-test}
```

That said, if I go for a specific attribute, it will fetch it from the cache:

```
nix-build -A rock64 -E 'with import <nixpkgs> {}; pkgsCross.aarch64-multiplatform.callPackage ./default.nix {}'
copying path '/nix/store/jn90vz3rh9dcp3bbqlwnycrmmfchl35y-util-linux-2.36.1-dev' from 'https://cache.nixos.org'...
building '/nix/store/cn9...-nixos-sd-image-20.09.2132.999b9b7db20-aarch64-linux.img.zst.drv'...

trying https://hydra.nixos.org/build/132043163/download/1/nixos-sd-image-20.09.2132.999b9b7db20-aarch64-linux.img.zst
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
```

And it finishes successfully:

```
No errors detected.
Header version: 1.0
Using 3 out of 128 partitions.
A total of 0 free sectors is available in 0 segments (the largest is (null)).
post-installation fixup
/nix/store/v8pmyvwivp5yapnjr0kwmpr0bqfizggb-image-aarch64-unknown-linux-gnu
```

I'll try to build with my changes later today and report back.

**Comment by Mic92** [comment]
> Got it. It looks like it still does not want to build (fails on arm-trusted-firmware):
> 
> ```
> # nix-build -E 'with import <nixpkgs> {}; pkgsCross.aarch64-multiplatform.callPackage ./default.nix {}'
> ...
> error: a 'aarch64-linux' with features {} is required to build '/nix/store/mr166g96ync4hlnm4046y878jnp1yyv0-arm-trusted-firmware-rk3399-2.4.drv', but I am a 'x86_64-linux' with features {benchmark, big-parallel, nixos-test}
> ```

What is the derivation for that?

**Comment by tmountain** [comment]
Glad to see that the NanoPC-T4 uboot PR got merged. I will circle back with a tested PR for the nixos-aarch64-images once the hydra builds get caught up with master. Thanks!
